guardian/newsletters-nx
# Data transformation

## What does this change?

Outlines a system for transforming our TBC data model to the legacy API format in the data-client library.
 - a new data model can be defined and edited as a `zodSchema` and the type and validation functions generated from the schema
 - The `deriveNewsletter` function has to be updated to take account of any changes to the `newsletterDataSchema`
 - A set of unit test can verify that the transformation module works as expected.

## How to test
`nx run newsletters-data-client:test --testFile=transform` will run the unit tests (they should pass)

Making changes to `newsletterDataSchema`, such as removing a field  required on the `Newsletter` type should make the test fail, until the `deriveNewsletter` function is updated to ensure the field is added as required.

E.G.  try removing  `listIdV1: z.number(),` from the `newsletterDataSchema` and the test will fail



## How can we measure success?

It should be possible to define our new data model using a `zodSchema` and convert it safely to the legacy API format. 

*update* - the advantage of having this in the codebase before we know what our data model will be is that it should allow us to make sure that it remains possible to derive the old data from the new data model as we develop it - for example, if we forget that we need to keep the `listId` field for backwards compatability, we should get a test error.

## Have we considered potential risks?

`newsletterDataSchema` is just a placeholder for the new data model. There are comments to indicate this and the type and functions are not exported by the data client library.

We will need to expand on the unit test as we go to ensure that the converted data has the correct values (not just the correct type).
